### PR TITLE
stream nar without nix-store --dump

### LINKED ISSUE #31: In-memory NarStore is unsound
Currently nar hashes are stored in memory, requiring the client to request the nar info before the nar file.
However this will break if the client has already the nar info is cached an harmonia was restarted after the fact.
In this case harmonia will return a 404.
Also this NarStore leaks memory over time.

### PR DESCRIPTION
**Pull request by Mic92**
Actually need to do the benchmark of removing `nix-store --dump`.

fix #31

### CODE REVIEW

**Comment by Conni2461** [comment]
Thanks for opening a PR :)

I gonna do a little bit of performance testing before merging but i think this will be performance improvement.

File: libnixstore/src/lib.rs [hunk 1]
```diff
@@ -339,3 +341,20 @@ pub fn get_build_log(derivation_path: &str) -> Option<String> {
...
+fn dump_path_trampoline<F>(data: &[u8], userdata: usize) -> bool
+where
+    F: FnMut(&[u8]) -> bool,
+{
+    let closure = unsafe { &mut *((userdata as *mut std::ffi::c_void) as *mut F) };
+    closure(data)
+}
+
+#[inline]
+pub fn dump_path<F>(store_path: &str, callback: F)
```
Comment (Conni2461) [inline libnixstore/src/lib.rs:354]: We need docs here, once everything is done and works

**Comment by Mic92** [comment]
> ```
> // nix-store --dump
> real    0m1.337s
> user    0m0.122s
> sys     0m0.255s
> 
> // pr
> real    0m0.929s
> user    0m0.075s
> sys     0m0.102s
> ```
> 
> But it doesnt start immediately streaming with this PR, but that isnt such a big deal. Tested with `/nix/store/7rl3p47k0zz4gczwsswg0jc887izm9bk-openjdk-17.0.3+7`. ~@ajs124 should probably do some tests on our aarch hardware.~, nvm i gonna test it on hb04

The speedup should be bigger on smaller nars than openjdk.

File: libnixstore/src/lib.rs (hunk 1 above)
Comment (Mic92) [reply libnixstore/src/lib.rs:355]: Fixed.

**Comment by Conni2461** [comment]
Thanks again :)

**Pull request merged**